### PR TITLE
Add simple frictional inelastic ball-ball collision model

### PR DESCRIPTION
**Pull request by derek-mcblane**
This model includes ball-ball inelasticity and ball-ball friction.  It does not include ball-table friction. Unlike the Mathavan ball-ball model, this does not require any numerical integration; everything is calculated from an impulse perspective, using a constant ball-ball coefficient of restitution for the velocities normal to the contact surfaces.  This comes at the cost of being less realistic.

I make no guarantees about the correctness of the model.  It looks more-or-less correct to me visually, and it passes the tests, but I haven't checked it thoroughly. I plan to write up my work so others can take a look.

**Some Background:**
This model is largely inspired by the analyses done by Dr. David Alciatore in his [technical proofs](https://billiards.colostate.edu/technical-proof/). In particular:
- [TP A-5](https://billiards.colostate.edu/technical_proofs/new/TP_A-5.pdf)
- [TP A-6](https://billiards.colostate.edu/technical_proofs/new/TP_A-6.pdf)
- [TP A-14](https://billiards.colostate.edu/technical_proofs/new/TP_A-14.pdf)

I also considered Marlow's "The General Collision" from "The Physics of Pocket Billiards". His model does account for ball-table friction, but it doesn't appear to account for inelasticity.  Ultimately, I didn't use anything from those models.

I haven't done a thorough search, but I wouldn't be surprised if this exact model has been derived before. I extended the ideas presented in those technical proofs to include motion of both balls.  I also found the complete final velocities and angular velocities of both balls when accounting for ball-ball inelasticity, ball-ball friction, and whether or not the balls finish the collision with any relative surface velocity.

**Future Work:**
I would like to add speed-dependent ball-ball coefficient of sliding friction like what Dr. Dave did in TP A-14. I haven't done so because it requires assuming that the velocity remains constant during the collision, which it obviously does not (although it might be a reasonable enough assumption).  I'll have to take a more thorough look at the Marlow data referenced in TP_A-14 to see under what conditions it was collected.

### CODE REVIEW

**Comment by ekiefl** [comment]
This is *awesome*! I took a quick look and it seems great. I'll properly review this weekend.

**Review by ekiefl — APPROVED** [review]
This is really cool. I also wonder whether this has been derived before... Maybe Dave would be interested in a new TP.

I made some additions to the test suite.

I was able to include the Mathavan model in your inelasticity test but I had to increase the number of iterations (https://github.com/ekiefl/pooltool/pull/155/commits/4682efe8a73a563e1efaae202a9e5c66a30ed2ab)

I extended the z-spin test case by parameterizing the magnitude of spin (https://github.com/ekiefl/pooltool/pull/155/commits/5546c94d04a62ffcf3025d074b24ab05084adc06).

In the head on inelastic test I parameterized coeff of restitution and added a check where the final cue ball has positive final velocity except for the case where e_b == 1 (perfect elasticity) (https://github.com/ekiefl/pooltool/pull/155/commits/919b785919c932fd2d39c97204a914ff4463ccdd)

Feel free to merge when you're ready.

File: pooltool/ptmath/utils.py [hunk 1]
```diff
@@ -257,14 +257,22 @@ def norm2d(vec: NDArray[np.float64]) -> float:
...
+) -> NDArray[np.float64]:
+    """Compute velocity of a point on ball's surface (specified by unit direction vector)"""
+    _, v, w = rvw
+    return v + cross(w, R * d)
+
+
 @jit(nopython=True, cache=const.use_numba_cache)
 def rel_velocity(rvw: NDArray[np.float64], R: float) -> NDArray[np.float64]:
-    """Compute velocity of cloth with respect to ball's point of contact
+    """Compute velocity of ball's point of contact with the cloth relative to the cloth
```
[the remark this answers is not in the document]
Comment (derek-mcblane) [reply pooltool/ptmath/utils.py:271]: Thanks.  I also changed the comment of `rel_velocity`.  I think this is more correct.

**Comment by derek-mcblane** [comment]
> This is really cool. I also wonder whether this has been derived before... Maybe Dave would be interested in a new TP.

Thanks! I'm writing up my work -- I'll share it soon.

> I made some additions to the test suite.
> 
> I was able to include the Mathavan model in your inelasticity test but I had to increase the number of iterations ([4682efe](https://github.com/ekiefl/pooltool/commit/4682efe8a73a563e1efaae202a9e5c66a30ed2ab))
> 
> I extended the z-spin test case by parameterizing the magnitude of spin ([5546c94](https://github.com/ekiefl/pooltool/commit/5546c94d04a62ffcf3025d074b24ab05084adc06)).
> 
> In the head on inelastic test I parameterized coeff of restitution and added a check where the final cue ball has positive final velocity except for the case where e_b == 1 (perfect elasticity) ([919b785](https://github.com/ekiefl/pooltool/commit/919b785919c932fd2d39c97204a914ff4463ccdd))

Looks good, thanks for cleaning this up.

> ~Feel free to merge when you're ready.~ You might not be able to merge due to branch protections, but let me know when you're ready and I'll merge it!

I think I'll add speed-dependent friction in a separate PR.  I want to check a couple things, and then it should be good to go.

**Comment by derek-mcblane** [comment]
Okay, I think this is good to go.

**Comment by ekiefl** [comment]
Merged!

Thank you for your important contribution :) I hope you consider contributing in the future.

**Pull request merged**